README.md

```diff
@@ -93,13 +93,40 @@ subscriptions model =
 
 #### Use it
 
-To show a message popup simply call one of the 4 type functions (`danger`, `info`, `success`, `warning`) for the defined `MessageToast` and update it to the model.
+To show a message popup simply generate a toast with one of the 4 type functions (`danger`, `info`, `success`, `warning`) and call a function onto it to display it with your individual needs.
+The simplest way is done by piping the generated toast into `MessageToast.withMessage`.
+
+**Example:**
+
+```elm
+-- Message that assigns a new "Danger" message toast to the MessageToast handler with default HTML layout.
+
+ShowDanger ->
+    let
+        messageToast =
+            model.messageToast
+                |> MessageToast.warning
+                |> MessageToast.withMessage "My warning message"
+    in
+    ( { model | messageToast = messageToast }, Cmd.none)
+```
+
+This will display the message with the default HTML layout of the MessageToast.
+If you want to customize the layout, you can instead pipe the generated toast into `MessageToast.withHtml` and provide a individual structured HTML layout for the content of the MessageToast.
+
+**Example:**
 
 ```elm
--- Message that assigns a new "Danger" message to the MessageToast handler.
+-- Message that assigns a new "Danger" message toast to the MessageToast handler with user defined HTML layout.
 
 ShowDanger ->
-    ( { model | messageToast = MessageToast.danger model.messageToast "Something critical happened." }, Cmd.none )
+    let
+        messageToast =
+            model.messageToast
+                |> MessageToast.warning
```

Comment:
I think it would be nicer to just introduce new functions for the html case. Because than we don't complicate the easy case and the default case should always be very easy. Additionally with this version one could forget to pass in the message at all and I don't like that.

So I would just add functions like:

```elm
MessageToast.htmlWarning
MessageToast.htmlDanger
...
```

Reply:
The question is what happens when we add more features later that will change the toast to any individual needs (`MessageToast.withCustomIcon`, `MessageToast.withReference`, ...).

I did ittthis way because I think it's still pretty easy to use if you want the default behaviour. But can also easily  change the toast flexibly to your needs. Basically it looks like:

```elm
messageToast
    |> MessageToast.apply <opt>
    |> MessageToast.apply <opt>
    |> MessageToast.apply <opt>
    |> MessageToast.construct
```

Where the `withMessage` or `withHtml` function are the ones that construct. Perhaps a slightly different naming can make it more clear, that this is the final step.

I don't think that the default case got more complicated. All you have to do now is write "MessageToast.withMessage" instead of passing the string directly. But it got way more explicit now IMO.

Reply:
I second @mschindlerMM. It is pretty explicit now while still being quite flexible.

I also like that there is a differentiation between the kind of toast you are displaying (i.e. warning) and how you decide to display it (i.e. withMessage ...).

Reply:
I don't like to change the API because we _might_ implement features that would make it nicer. 

And even with that in mind I think it works pretty well. We would have:

```elm
 MessageToast.warning model.messageToast "My warning message"
   |> MessageToast.withCustomIcon "bla"

-- And the fully customized version:

 MessageToast.htmlWarning model.messageToast (div [] [text "My warning message"])
```

This has the advantage that the first use and default use case is still super easy while still providing all the options for customization. I also like that the fully customized version is different than the one that changes only a single thing. Because this is not really compatible. You can not pass in a fully customized one and change the icon for example.

Reply:
> ```elm
>  MessageToast.warning model.messageToast "My warning message"
>    |> MessageToast.withCustomIcon "bla"
> ```

This wouldn't work as nicely. `warning` would already put the toast into the list of all toasts. Therefore we would need to pull out the last toast in order to apply the `withCustomIcon` function.. which could maybe lead to problems (race-conditions). I'm not fully sure if it could end up in a race-condition, but it is still not as nicely coupled.
 
> Because this is not really compatible. You can not pass in a fully customized one and change the icon for example.

This is not true. At least it's not how it is implemented right now and also not intended. The goal of the HTML function is not to re-define the whole toast layout yourself. It is used to structure the content of the toast individually (like having a title, button etc. inside), instead of plain text.

```elm
-- Instead of:
text "My toast message"

-- The user can define:
div [] [ text "Toast title, text "Toast body" ]
```

> I don't like to change the API because we might implement features

I honestly rather break the API right now, knowing that the change will make it better to maintain in the future. Also since the breaking change is easy to fix and will not make it any verbose, but make it more explicit.

I'm currently still not convinced by your idea. But as I might have already said, I'm fine with renaming the `withHtml` and `withMessage` functions, to make it more clear, that they are the final constructing step..

Reply:
There would not be a race condition because javascript can not execute other code while it is executing this.But I misunderstood this. Changing one that is already added is not nice.

I still think we should make the easy use case very easy and without needing to call multiple functions. So maybe we should add functions to create and add in one step and another to just create one that can be modified and later added. I just think adding the ability to customize should never complicate first use and simple use.

Reply:
I just don't really like having different functions for a "simple" toast and a "customizable" toast. Sometimes it makes totally sense for a package to differentiate between a fully pre-configured version and a version you can fully customize on you own.

In this case, however, it would be better if you don't need to worry about a special created toast, in order to apply some custom changes. You just have your toast and adjust it or not ... however you wish. It's like the FeatherIcon's way... 

```elm
-- easy case:
FeatherIcons.x
    |> FeatherIcons.toHtml []

-- custom case:
FeatherIcons.x
    |> FeatherIcons.withSize 14
    |> FeatherIcons.toHtml []
```

Especially because the easy case didn't even got more complicated .. but I already said that. 

I feel like that this discussion is stagnating as no one seems to be fully convinced yet about each others idea.
Maybe we should just do a little survey and ask each developer on the team to give a vote for either alternative (or they find the a better alternative)... so that we can solve the issue at some point 😄

Reply:
After thinking more about it and reading your latest reply I am convinced now. It always makes sense to differentiate between creating and adding it. It should not work completely different.